### ISSUE TITLE
feature request: list of leaf indices

### ISSUE THREAD

**Issue by Far0n** [dmlc/xgboost#105]
hello together,

it would be nice to be able to use xgboost for feature transformation. That is, the abilitiy to get a list/matrix of leaf indices instead of predictions (similar to http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.apply)

best regards

**Comment by leakyMirror** [comment]
Is it possible to do the same with command line tool?

**Comment by tqchen** [comment]
It is not supported now, for now. You can do this simply with python by load the model trained from command line, and DMatrix from the libsvm file and call predict with pred_leaf

**Comment by binga** [comment]
Hello,

I have a dataset with 9 distinct levels in the target. I would like to get the leaf index of each data point in say, 20 trees. My understanding is that, for each target level, the data points might end up at different leaf index. 

dtrain = xgb.DMatrix(train, label = y)
param = {'objective':'multi:softprob', 'eval_metric':'mlogloss', 'num_class':9, 'silent':1, 'booster':'gbtree'}

num_rounds = 20
bst = xgb.train(param, dtrain, num_rounds)

train_leaf_preds = bst.predict(dtrain, pred_leaf = True)

By doing print train_leaf_preds.shape, I get ndata x 180(20trees * 9 target levels) features ndarray.

How do I interpret this? Is it that, the first 20 columns correspond to leaf indices when the first target level is considered and so on..?

Thank you for the awesome library Tianqi! It's incredible :-)

**Comment by antant-shenzhen-2013** [comment]
@tqchen as you said before,

> It is not supported now, for now. You can do this simply with python by load the model trained from command line, and DMatrix from the libsvm file and call predict with pred_leaf

if number of samples to predict is too big to be hold in one machine memory, does your suggestion still work?  Seems still need command line support for large dataset that need feature transformation.

**Comment by antant-shenzhen-2013** [comment]
follow this idea, can be done.
https://aws.amazon.com/blogs/machine-learning/distributed-inference-using-apache-mxnet-and-apache-spark-on-amazon-emr/
